odoo/addons/base/models/ir_asset.py

```diff
@@ -0,0 +1,334 @@
+        certain bundle. The returned list is composed of tuples containing the
+        file path [1] and the first addon calling it [0]. Asset loading is
+        performed as following:
+
+        1. At each initial call (i.e. not a recursive call), a new list of
+        assets is generated.
+
+        2. All 'ir.asset' records matching the given bundle and with a sequence
+        number inferior to 16 are applied.
+
+        3. The manifests of all given addons are checked for assets declaration
+        for the given bundle. If any, they are read sequentially and their
+        operations are applied to the current list.
+
+        4. After all manifests have been parsed, the remaining 'ir.asset'
+        records matching the bundle are also applied to the current list.
+
+        :param addons: list of addon names as strings. The files returned will
+            only be contained in the given addons.
+        :param bundle: name of the bundle from which to fetch the file paths
+        :param css: boolean: whether or not to include style files
+        :param js: boolean: whether or not to include script files
+        :param xml: boolean: whether or not to include template files
+        :param asset_paths: (addon, path)[]: the current list of loaded assets.
+            It starts blank (initial) and is given to each subsequent call.
+        :returns: the list of tuples (addon, file_path)
+        """
+        exts = []
+        manifest_cache = self._get_manifest_cache()
+        if js:
+            exts += SCRIPT_EXTENSIONS
+        if css:
+            exts += STYLE_EXTENSIONS
+        if xml:
+            exts += TEMPLATE_EXTENSIONS
+
+        if addons is None:
+            addons = self._get_addons_list()
+
+        # 1. Creates an empty assets list (if initial call).
+        if asset_paths is None:
+            asset_paths = []
+        # This index will be used when prepending: files will be prepend at the
+        # start of the CURRENT bundle.
+        bundle_start_index = len(asset_paths)
+
+        def get_path_index(path, raise_if_none=True):
+            """Returns the index of the given path in the current assets list."""
+            for i in range(len(asset_paths)):
+                if asset_paths[i][0] == path:
+                    return i
+            if raise_if_none:
+                raise Exception("File %s not found in bundle %s" % (path, bundle))
+
+        def add_paths(addon, paths, target_index=None):
+            """Adds the given paths to the current list. An index can be specified."""
+            target_index = len(asset_paths) if target_index is None else target_index
+            for path in paths:
+                if get_path_index(path, False) is not None:
+                    # the path is already present in the asset list: don't duplicate it
+                    continue
+                asset_paths.insert(target_index, (path, addon, bundle))
+                target_index += 1
```

Comment:
This for loop has N² complexity.
You can check presence in O(1) if you maintain a `set` with the paths in `asset_paths`.